### PR TITLE
Improves viewers and oviewers.

### LINKED ISSUE #1586: Spontaneous gib
Ike said it was probably because a lavaland mob gibbed me across z levels

On Paradise SS14 test

### LINKED ISSUE #1589: Minebot tries to attack slimes across zlevels
image

Minebot on lavaland, slime on the station in xenobio.

Need to dig into the associated DM code.

### PR DESCRIPTION
**Pull request by Laboredih123**
Hoorah, we now do opaque blocking and also check on only our zlevel.
Fixes #1586
Fixes #1589

### CODE REVIEW

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 1]
```diff
@@ -3049,22 +3063,37 @@ public static DreamValue NativeProc_viewers(NativeProc.Bundle bundle, DreamObjec
...
-                if (Math.Abs(centerX - mob.X) <= depth && Math.Abs(centerY - mob.Y) <= depth) {
-                    view.AddValue(new DreamValue(mob));
+                if (Math.Abs(centerPos.X - mob.X) <= depth && Math.Abs(centerPos.Y - mob.Y) <= depth && centerPos.Z == mob.Z) {
+                    (_, ViewRange range) = DreamProcNativeHelpers.ResolveViewArguments(bundle.DreamManager, mob, bundle.Arguments);
+                    var eyePos = bundle.AtomManager.GetAtomPosition(mob);
+                    var viewData = DreamProcNativeHelpers.CollectViewData(bundle.AtomManager, bundle.MapManager, eyePos, range);
+
+                    ViewAlgorithm.CalculateVisibility(viewData);
+
+                    foreach (var tile in DreamProcNativeHelpers.MakeViewSpiral(viewData, true)) {
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:3086]: According to this, `viewers()` does not go in an outward spiral like `view()` does. It instead goes by column.
image

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs [hunk 2]
```diff
@@ -1855,8 +1855,24 @@ public static DreamValue NativeProc_oviewers(NativeProc.Bundle bundle, DreamObje
...
-                if (Math.Abs(centerPos.X - mob.X) <= depth && Math.Abs(centerPos.Y - mob.Y) <= depth) {
-                    view.AddValue(new DreamValue(mob));
+                if (Math.Abs(centerPos.X - mob.X) <= depth && Math.Abs(centerPos.Y - mob.Y) <= depth && centerPos.Z == mob.Z) {
+                    (_, ViewRange range) = DreamProcNativeHelpers.ResolveViewArguments(bundle.DreamManager, mob, bundle.Arguments);
+                    var eyePos = bundle.AtomManager.GetAtomPosition(mob);
+                    var viewData = DreamProcNativeHelpers.CollectViewData(bundle.AtomManager, bundle.MapManager, eyePos, range);
+
+                    ViewAlgorithm.CalculateVisibility(viewData);
+
+                    foreach (var tile in DreamProcNativeHelpers.MakeViewSpiral(viewData, true)) {
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:1865]: This also goes by column, like `viewers()`.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs (hunk 1 above)
Comment (ZeWaka) [reply OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:3086]: lol

Comment (Laboredih123) [reply OpenDreamRuntime/Procs/Native/DreamProcNativeRoot.cs:3086]: lmao

**Comment by wixoaGit** [comment]
This doesn't replicate the column-by-column result in BYOND but that can be a rework for another day if it ends up being  an issue.

**Review by wixoaGit — APPROVED** [review]

**Comment by Laboredih123** [comment]
observe as a codebase explodes

**Pull request merged**